base/utf32.jl

```diff
@@ -20,15 +20,234 @@ length(s::UTF32String) = length(s.data) - 1
+        @inbounds buf[1] = Char(0)
+        return UTF32String(buf)
+    end
+    # get number of words to create
+    len, cnt2::UInt, cnt3::UInt, cnt4::UInt, flags::UInt = utf8_validate(iStr)
+    buf = Array{Char,1}(len+1)
+    @inbounds begin
+    buf[len+1] = Char(0) # NULL termination
+    if flags == 0
+        # 7-bit ASCII characters only! Do optimized conversion
+        for pos=1:len
+            buf[pos] = Char(iStr[pos])
+        end
+    else
+        # has multi-byte UTF-8 sequences
+        out::UInt = 0
+        pos::UInt = 0
+        while out < len
+            ch::UInt32 = iStr[pos += 1]
+            if ch > 0x7f
+                if ch < 0xe0
+                    ch = ((ch & 0x1f) << 6) | (iStr[pos += 1] & 0x3f)
+                elseif ch < 0xf0
+                    ch = ((ch & 0xf) << 12) | (UInt32(iStr[pos+1] & 0x3f) << 6) | (iStr[pos+2] & 0x3f)
+                    pos += 2
+		    if is_surrogate_lead(ch)
+			ch = (((ch & 0x3ff) << 10)
+				| (((UInt32(iStr[pos+1] & 0xf) << 12)
+				|  (UInt32(iStr[pos+2] & 0x3f) << 6)
+				|  (iStr[pos+3] & 0x3f)) & 0x3ff)) + 0x10000
+			pos += 3
+		    end
+                else
+                    ch = ((ch & 0x7) << 18) | (UInt32(iStr[pos+1] & 0x3f) << 12) | (UInt32(iStr[pos+2] & 0x3f) << 6) | (iStr[pos+3] & 0x3f)
+                    pos += 3
+                end
+            end
+            buf[out += 1] = Char(ch)
+        end
+    end
+    end
+    UTF32String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF-16 encoded string to UTF-32 encoding
+
+@param[in]  iStr::UTF16String
+
+@return     ::UTF32String
+@throws     UnicodeError
+"""
+=#
+function utf16_to_utf32(iStr::Array{UInt16,1})
+    len::UInt = (sizeof(iStr) >>> 1) - 1 # account for trailing \0
+    # handle zero length string quickly
+    if len == 0
+        buf = Array{Char,1}(1)
+        @inbounds buf[1] = Char(0)
+        return UTF32String(buf)
+    end
+    # get number of words to create
+    len, cnt2::UInt, cnt3::UInt, cnt4::UInt, flags::UInt = utf16_validate(iStr)
+    buf = Array{Char,1}(len+1)
+    @inbounds buf[len+1] = Char(0) # NULL termination
+    if (flags & UTF_UNICODE4) == 0
+        # No surrogate pairs, do optimized copy
+        for pos=1:len
+            @inbounds buf[pos] = Char(iStr[pos])
+        end
+    else
+        # has multi-byte UTF-8 sequences
+        out::UInt = 0
+        pos::UInt = 0
+        while out < len
+            @inbounds ch::UInt32 = iStr[pos += 1]
+            # check for surrogate pair
+            if is_surrogate_lead(ch)
+                @inbounds ch = 0x10000 + (((ch & 0x3ff) << 10) | (iStr[pos += 1] & 0x3ff))
+            end
+            @inbounds buf[out += 1] = Char(ch)
+        end
+    end
+    UTF32String(buf)
+end
+
+#=
+"""
+@brief      Converts a UTF-32 encoded string to UTF-16 encoding
+
+@param[in]  iStr::UTF32String
+
+@return     ::UTF16String
+@throws     UnicodeError
+"""
+=#
+function utf32_to_utf16(iStr::Array{Char,1})
```

Comment:
I'd have to have a different one for `AbstractString`, that did `for ch in s` instead of a `while` loop, but is otherwise the same as `utf32_validate`.